modules/data-table/main/index.coffee

```diff
@@ -59,100 +265,182 @@ class DataTable extends hx.EventEmitter
+
+    clearFilters = => @advancedSearch(undefined)
+
+    advancedSearchAddFilterButton = hx.button({context: 'positive'})
+      .classed('hx-data-table-advanced-search-button hx-btn-invert', true)
+      .add(hx.icon({class: 'hx-data-table-advanced-search-icon hx-icon hx-icon-plus hx-text-positive'}))
+      .add(hx.detached('span').text(resolvedOptions.addFilterText))
+      .on 'click', addFilter
+
+    advancedSearchClearFilterButton = hx.button({context: 'negative'})
+      .classed('hx-data-table-advanced-search-clear-filters hx-data-table-advanced-search-button hx-btn-invert', true)
+      .add(hx.icon({class: 'hx-data-table-advanced-search-icon hx-icon hx-icon-close hx-text-negative'}))
+      .add(hx.detached('span').text(resolvedOptions.clearFiltersText))
+      .on 'click', clearFilters
+
+    # We create multiple copies of these to show in different places
+    # This makes it easier to change the UI as we can show/hide instead of moving them
+    [pageSize, pageSizePicker] = createPageSizeBlock(this, resolvedOptions)
+    [pageSizeBottom, pageSizePickerBottom] = createPageSizeBlock(this, resolvedOptions)
+
+    [pagination, pagePicker] = createPaginationBlock(this)
+    [paginationBottom, pagePickerBottom] = createPaginationBlock(this)
+    [paginationCompact, pagePickerCompact] = createPaginationBlock(this)
+
+    # The main pagination is hidden as the compact control panel contains a version of it
+    pagination.classed('hx-data-table-compact-hide', true)
+
+    controlPanelBottom = hx.detached('div').class('hx-data-table-control-panel-bottom')
+
+    # Create the structure in one place
+    # Some entities still make sense to be built individually (e.g. the loading div)
+    selection
+      .add(content)
+      .add(statusBar
+        .add(statusBarText)
+        .add(statusBarClear))
+      # Control panel displayed at the top for compact mode
+      .add(controlPanelCompact
+        .add(paginationCompact)
+        .add(spacer())
+        .add(controlPanelCompactToggle))
+      # Main control panel - contains all the components
+      .add(controlPanel
+        .add(controlPanelInner
+          .add(compactSort)
+          .add(pagination)
+          .add(pageSize)
+          .add(spacer())
+          .add(filterContainer
+            .add(advancedSearchToggle)
+            .add(filterInput)))
+        # The advanced search container isn't in the main control panel as it is easier to style outside
+        .add(advancedSearchContainer
+          .add(advancedSearch)
+          .add(advancedSearchButtons
+            .add(advancedSearchAddFilterButton)
+            .add(advancedSearchClearFilterButton))))
+      # Bottom control panel - shown in compact mode and when the search is at the top
+      .add(controlPanelBottom
+        .add(spacer())
+        .add(pageSizeBottom)
+        .add(paginationBottom))
+      # Add the loading div last - helps keep it on top of everything
+      .add(loadingDiv)
```

Comment:
After scrolling up and down through the constructor repeatedly to work out how everything was constructed, I converted everything to detached divs and added the structure in one place - I think it's much easier to work out this way